core/optimization/src/main/java/it/unibz/inf/ontop/iq/optimizer/impl/PreventDistinctOptimizerImpl.java

```diff
@@ -0,0 +1,225 @@
+            super(optimizationSingletons.getCoreSingletons());
+            this.variableGenerator = variableGenerator;
+            this.termFactory = optimizationSingletons.getCoreSingletons().getTermFactory();
+            this.substitutionFactory = optimizationSingletons.getCoreSingletons().getSubstitutionFactory();
+        }
+
+        @Override
+        public IQTree transformConstruction(IQTree tree, ConstructionNode rootNode, IQTree child) {
+            var decomposition = Decomposition.decomposeTree(child);
+            if(decomposition.distinctNode.isPresent()) {
+                return performTransformation(rootNode, tree, decomposition);
+            }
+            return super.transformConstruction(tree, rootNode, child);
+        }
+
+        private IQTree performTransformation(ConstructionNode constructionNode, IQTree originalTree, Decomposition decomposition) {
+            var subtree = decomposition.descendantTree;
+
+
+            //Find terms that need to be pushed below the distinct.
+            ImmutableMap.Builder<Variable, ImmutableTerm> substitutionBuilder = ImmutableMap.builder();
+            var updatedTerms = constructionNode.getSubstitution().stream()
+                    .map(entry -> Maps.immutableEntry(entry.getKey(), splitByPreventDistinctType(entry.getValue(), substitutionBuilder)))
+                    .collect(ImmutableCollectors.toSet());
+            if(substitutionBuilder.build().isEmpty())
+                return originalTree;
+            var substitution = substitutionBuilder.build();
+            if(!substitution.values().stream()
+                    .allMatch(this::isDeterministic))
+                throw new IllegalArgumentException("Unable to push non-supported terms below DISTINCT, because they include a non-deterministic function.");
+
+
+            //Check if all variables used by these terms have a functional dependency pointing to them.
+            var pushedVariables = substitution.entrySet().stream()
+                    .flatMap(entry -> entry.getValue().getVariableStream())
+                    .collect(ImmutableCollectors.toSet());
+            var functionalDependencies = subtree.inferFunctionalDependencies();
+            if(pushedVariables.stream()
+                    .anyMatch(v -> functionalDependencies.getDeterminantsOf(v).stream()
+                            .noneMatch(determinants -> Sets.intersection(determinants, pushedVariables).isEmpty())))
+                throw new IllegalArgumentException(String.format("No Functional Dependency could be used to push the variables %s down.", pushedVariables));
+
+
+            //Construct the updated IQTree.
+            var topConstruction = iqFactory.createConstructionNode(
+                    constructionNode.getVariables(),
+                    updatedTerms.stream()
+                            .filter(entry -> !(entry.getValue() instanceof Variable))
```

Comment:
What if there were already variable-to-variable entries in the original substitution?